Add support for passing options to vega CLI utils via `vega_cli_options` kwarg

boydgreenfield opened the pull request:
My primary use case is to be able to render interactive charts with multiple renderers (`alt.renderers.enable("altair_saver", fmts=["html", "svg"]`) without getting warnings about the `window` coming back and getting rendered into a notebook or other `nbconvert` export. 

Closes #51. @jakevdp Would love your review and happy to tweak as needed – this was my quick and dirty approach to pass the arguments through, but if there's a more elegant way of doing so, please let me know!

jakevdp commented on altair_saver/_core.py:
```diff
@@ -160,9 +165,19 @@ def save(
         embed_options = alt.renderers.options.get("embed_options", None)
+    if vega_cli_options is None:
+        vega_cli_options = alt.renderers.options.get(
+            "embevega_cli_optionsd_options", None
```
Looks like a copy/paste error?

jakevdp commented on altair_saver/savers/_html.py:
```diff
@@ -122,6 +122,7 @@ def __init__(
         spec: JSONDict,
         mode: Optional[str] = None,
         embed_options: Optional[JSONDict] = None,
+        vega_cli_options: Optional[List] = None,
```
Strange that the vega_cli_options are passed here, when they're not used.

boydgreenfield replied:
The issue is that we need to pass `vega_cli_options` into all of the `Saver` methods. We can tweak that by instead updating `kwargs` manually:

```python3
if embed_options is None:
    embed_options = alt.renderers.options.get("embed_options", None)
if vega_cli_options is None:
    vega_cli_options = alt.renderers.options.get(
        "vega_cli_options", None
    )
kwargs["vega_cli_options"] = vega_cli_options

Saver = _select_saver(method, mode=mode, fmt=fmt, fp=fp)
saver = Saver(
    spec,
    mode=mode,
    embed_options=embed_options,
    **kwargs,
)
```

Preference?

boydgreenfield replied:
Note the above needs to check `issubclass(Saver, NodeSaver)` (or for equality). Updated as such in the 3rd commit on this PR.

boydgreenfield replied:
We're now just passing these through as `kwargs`. This does mean each `save()` invocation requires passing `vega_cli_optoins` if being called explicitly.

jakevdp commented on altair_saver/savers/_selenium.py:
```diff
@@ -160,6 +160,7 @@ def __init__(
         spec: JSONDict,
         mode: Optional[str] = None,
         embed_options: Optional[JSONDict] = None,
+        vega_cli_options: Optional[List] = None,
```
Unused here

boydgreenfield replied:
See comment above re: `HTMLSaver`. Either this needs to just be passed via the `kwargs` or if passed explicitly in the `__init__` for all the subclasses. 

Happy to refactor.

boydgreenfield replied:
This is refactored in the 3rd commit. Can be merged w/o that or not.

boydgreenfield replied:
This has been removed.

boydgreenfield commented:
> This looks good – a few comments.
> 
> One larger question: do the CLI commands all accept the same flags? Is a single list applied to all commands the best approach here?

They have the same interface, so I think it's worth keeping this simple? I don't _really_ think you'd want to vary any of the params per renderer except for maybe `--scale`:

These are the options for `vg2pdf`, `vg2svg`, and `vg2png`

```
Options:
  -b, --base        Base directory for data loading. Defaults to the directory
                    of the input spec.                                  [string]
  -l, --loglevel    Level of log messages written to stderr. One of "error",
                    "warn" (default), "info", or "debug".               [string]
  -c, --config      Vega config object. Either a JSON file or a .js file that
                    exports the config object.                          [string]
  -f, --format      Number format locale descriptor. Either a JSON file or a .js
                    file that exports the locale object.                [string]
  -t, --timeFormat  Date/time format locale descriptor. Either a JSON file or a
                    .js file that exports the locale object.            [string]
  -s, --scale       Output resolution scale factor.        [number] [default: 1]
  --seed            Seed for random number generation.                  [number]
  --help            Show help                                          [boolean]
  --version         Show version number                                [boolean]
```

boydgreenfield commented:
@jakevdp I fixed the typo (oops!) and added a commit to only pass `vega_cli_options` to `NodeSaver`. _But_ I can't reproduce the two failing tests on GH Actions locally (they pass on 3.7 for me). Maybe something about my renderers/environment config.

I don't really _love_ this solution though. It seems like it'd be cleaner to pass along to `NodeSaver` only and have a subclassed `__init__` method for it _or_ keep it at the `Saver` level and pass it down (which I did originally just because that's how `embed_options` works and that's actually only used in the 

 we should probably _only_ have it in `NodeSaver` or pass it through to everything. The former commit was just emulating the `embed_options` placements, which is also not used in all of the `Savers`.

jakevdp commented:
Thanks – I think we should avoid special-casing for subclasses within the base class. Probably a better approach is to let the subclasses accept arbitrary kwargs and ignore any unrecognized/unsupported ones.

boydgreenfield commented:
> Thanks – I think we should avoid special-casing for subclasses within the base class. Probably a better approach is to let the subclasses accept arbitrary kwargs and ignore any unrecognized/unsupported ones.

OK, I've refactored to have all of the subclasses take a `kwargs` and then this is only being added to (1) `_node.py` for `NodeSaver` and the related `vg2*` functions and documented in the `render` and `save` methods in `_core.py`.

Let me know if this looks good!

jakevdp commented:
Looks good! Last issue is to add some tests. I think we want two basic tests of this new behavior:

- a new test in test_node.py that ensures that keywords passed to NodeSaver are properly propagated to the CLI functions
- a new test in test_core.py that ensures that keywords passed to save() and render() make it to the NodeSaver

This will probably involve some mocking; if you're not familiar with mocking in pytest, search for ``monkeypatch`` within this repo to see a few examples of relevant approaches.

boydgreenfield commented:
> Looks good! Last issue is to add some tests. I think we want two basic tests of this new behavior:
> 
> * a new test in test_node.py that ensures that keywords passed to NodeSaver are properly propagated to the CLI functions
> * a new test in test_core.py that ensures that keywords passed to save() and render() make it to the NodeSaver
> 
> This will probably involve some mocking; if you're not familiar with mocking in pytest, search for `monkeypatch` within this repo to see a few examples of relevant approaches.

Added some tests. Tried to mimic the style and mocking set up you have already, but let me know if they need tweaks (and also do feel free to just tweak directly if easier). A couple of the tests felt a little awkward/verbose but needed extra runtime checks because of MyPy.

jakevdp commented:
Thanks - in the new test within ``test_core``, it doesn't appear that you're validating that the cli_options argument is being correctly passed to NodeSaver.

jakevdp commented on altair_saver/savers/_node.py:
```diff
@@ -34,33 +35,46 @@ def exec_path(name: str) -> str:
+    cmd = [vg2png]
+    if vega_cli_options:
+        cmd += vega_cli_options
+    return check_output_with_stderr(cmd, input=vg_json)
```
I'd make this a bit more concise:
```
return check_output_with_stderr([vg2png, *(vega_cli_options or [])], input=vg_json)
```
Similar in the other two functions

boydgreenfield replied:
Done.

boydgreenfield commented:
> Thanks - in the new test within `test_core`, it doesn't appear that you're validating that the cli_options argument is being correctly passed to NodeSaver: in other words, if `save()`/`render()` did nothing with that argument, the test would still pass.

These do shell out to `vg2*` – I've added a test to capture the stderr and ensure that, e.g., we get DEBUG logging when passing in the debug loglevel. 

I think this addresses all of your comments @jakevdp. Let me know if you need anything else to merge.

jakevdp commented on altair_saver/savers/tests/test_node.py:
```diff
@@ -37,9 +37,12 @@ def get_modes_and_formats() -> Iterator[Tuple[str, str]]:
+@pytest.mark.parametrize("mode,fmt", get_modes_and_formats())
+@pytest.mark.parametrize("vega_cli_options", [None, ["--loglevel", "error"]])
+def test_node_mimebundle(
+    name: str, data: Any, mode: str, fmt: str, vega_cli_options: Optional[List]
```
Optional[List[str]]

boydgreenfield replied:
Done.

jakevdp commented on altair_saver/tests/test_core.py:
```diff
@@ -257,6 +257,44 @@ def test_embed_options_save_html_override(spec: JSONDict) -> None:
+@pytest.mark.parametrize("fmt", ["html", "svg"])
+@pytest.mark.parametrize("vega_cli_options", [None, ["--loglevel", "debug"]])
+def test_save_w_vega_cli_options(
+    capsys: Any, chart: alt.TopLevelMixin, fmt: str, vega_cli_options: Optional[List]
```
capsys: SysCapture (use ``from _pytest.capture import SysCapture``)
vega_cli_options: Optional[List[str]]

boydgreenfield replied:
Done.

jakevdp commented on altair_saver/tests/test_core.py:
```diff
@@ -257,6 +257,44 @@ def test_embed_options_save_html_override(spec: JSONDict) -> None:
+
+@pytest.mark.parametrize("vega_cli_options", [None, ["--loglevel", "debug"]])
+def test_render_w_vega_cli_options(
+    capsys: Any, chart: alt.TopLevelMixin, vega_cli_options: Optional[List]
```
types, same as above.

boydgreenfield replied:
Done.

jakevdp commented on altair_saver/tests/test_core.py:
```diff
@@ -257,6 +257,44 @@ def test_embed_options_save_html_override(spec: JSONDict) -> None:
+    bundle = render(chart, fmts=["html", "svg"], vega_cli_options=vega_cli_options)
+    assert len(bundle) == 2
+    for mimetype, content in bundle.items():
+        fmt = mimetype_to_fmt(mimetype)
```
assert content is not None (for mypy)

boydgreenfield replied:
Done.

boydgreenfield commented on altair_saver/_core.py:
```diff
@@ -162,7 +165,7 @@ def save(
 
     Saver = _select_saver(method, mode=mode, fmt=fmt, fp=fp)
-    saver = Saver(spec, mode=mode, embed_options=embed_options, **kwargs)
+    saver = Saver(spec, mode=mode, embed_options=embed_options, **kwargs,)
```
Done.

boydgreenfield commented on altair_saver/_core.py:
```diff
@@ -242,7 +248,7 @@ def render(
     for fmt in fmts:
         Saver = _select_saver(method, mode=mode, fmt=fmt)
-        saver = Saver(spec, mode=mode, embed_options=embed_options, **kwargs)
+        saver = Saver(spec, mode=mode, embed_options=embed_options, **kwargs,)
```
Done.

boydgreenfield commented on altair_saver/savers/_node.py:
```diff
@@ -69,6 +72,28 @@ class NodeSaver(Saver):
         "vega": ["pdf", "png", "svg"],
         "vega-lite": ["pdf", "png", "svg", "vega"],
     }
+    _vega_cli_options: List
```
Done.

boydgreenfield commented on altair_saver/savers/_node.py:
```diff
@@ -69,6 +72,28 @@ class NodeSaver(Saver):
+        embed_options: Optional[JSONDict] = None,
+        vega_version: str = alt.VEGA_VERSION,
+        vegalite_version: str = alt.VEGALITE_VERSION,
+        vegaembed_version: str = alt.VEGAEMBED_VERSION,
```
Done.

boydgreenfield commented on altair_saver/savers/_node.py:
```diff
@@ -69,6 +72,28 @@ class NodeSaver(Saver):
+            embed_options=embed_options,
+            vega_version=vega_version,
+            vegalite_version=vegalite_version,
+            vegaembed_version=vegaembed_version,
```
Done.

boydgreenfield commented on altair_saver/savers/_selenium.py:
```diff
@@ -167,6 +167,7 @@ def __init__(
         webdriver: Optional[Union[str, WebDriver]] = None,
         offline: bool = True,
         scale_factor: Optional[float] = 1,
+        **kwargs,
```
Done.

boydgreenfield commented:
@jakevdp All the above comments have been addressed and tests are passing. I think between this and #53, most of issues problems highlighted by #52 should be addressed or solvable via passing CLI args. Thanks again!

jakevdp commented:
Awesome, thanks for the contribution!
